README.md
# EngineCL

Usability and Performance in Heterogeneous Computing.

## Core

The core is extracted as a library, exposing the Tier-2.

### Targets

Building targets for debug, release or debug-test.

```
make build/debug
make build/release
make build/debug-test
```

### Examples

EngineCL-tier2 is provided to be able to execute Saxpy benchmark. It can use two load balancing algorithms.

#### Static

Static algorithm with devices 0.0 and 1.0 (platform.device), problem size of 1024, chunksize of 128, results are checked. The work is divided in half (0.5 is given to first device).

The static props are the proportions given to the devices. Eg. if using 3 devices, we can provide 2 or 3 static props. Therefore, `0.5:0.2` or `0.5:0.2:0.3` are the same: 50% for the first, 20% for the second, 30% (rest) for the third.

```
 ❯ ./build/debug/EngineCL-tier2 1024 128 3.14 --static 0.5 --check --devices 0.0,1.0
Config:
  scheduler: static
  size: 1024
  chunksize: 128
  constant: 3.14
  check: yes
  kernel path:examples/tier-2/saxpy.cl
  platform.device list: 0.0 1.0
  static props: 0.5
  dynamic chunks: 1
time: 510
Kernel: saxpy
Runtime init timestamp: 1550921407
Runtime durations:
 initDiscovery: 117 ms.
Device id: 0
Selected platform: Clover
Selected device: AMD Radeon (TM) R7 300 Series (PITCAIRN, DRM 3.27.0, 4.20.10-arch1-1-ARCH, LLVM 7.0.1)
program type: source
kernel: saxpy
works: 1 works_size: 512
duration increments:
 init: 0 ms.
 useDiscovery: 0 ms.
 initContext: 0 ms.
 initQueue: 3 ms.
 initBuffers: 0 ms.
 initKernel: 387 ms.
 writeBuffers: 0 ms.
 deviceStart: 0 ms.
 deviceReady: 0 ms.
 deviceRun: 0 ms.
 deviceEnd: 0 ms.
 completeWork: 0 ms.
 total: 390 ms.
duration offsets from init:
 init: 118 ms.
 useDiscovery: 118 ms.
 initBuffers: 122 ms.
 initKernel: 509 ms.
 writeBuffers: 509 ms.
 deviceStart: 509 ms.
 deviceReady: 509 ms.
 deviceRun: 509 ms.
 deviceEnd: 510 ms.
use events: yes
chunks (mOffset+mSize:ts_ms+duration_ms)type-chunks,0+512:509+1,
Device id: 1
Selected platform: AMD Accelerated Parallel Processing
Selected device: AMD FX-8320E Eight-Core Processor
program type: source
kernel: saxpy
works: 1 works_size: 512
duration increments:
 init: 0 ms.
 useDiscovery: 0 ms.
 initContext: 0 ms.
 initQueue: 2 ms.
 initBuffers: 0 ms.
 initKernel: 40 ms.
 writeBuffers: 0 ms.
 deviceStart: 0 ms.
 deviceReady: 0 ms.
 deviceRun: 0 ms.
 deviceEnd: 0 ms.
 completeWork: 0 ms.
 total: 42 ms.
duration offsets from init:
 init: 118 ms.
 useDiscovery: 118 ms.
 initBuffers: 120 ms.
 initKernel: 161 ms.
 writeBuffers: 161 ms.
 deviceStart: 161 ms.
 deviceReady: 161 ms.
 deviceRun: 161 ms.
 deviceEnd: 161 ms.
use events: yes
chunks (mOffset+mSize:ts_ms+duration_ms)type-chunks,512+512:161+0,
StaticScheduler:
chunks: 2
duration offsets from init:
 schedulerStart: 1 ms.
 schedulerEnd: 510 ms.
Success
```

### Dynamic

Dynamic algorithm with devices 0.0 and 1.0 (platform.device), problem size of 102400000, chunksize of 128, results are checked. The work is divided dynamically in 8 chunks, each one given to the first device that is free.

```
 ❯ ./build/debug/EngineCL-tier2 102400000 128 3.14 --dynamic 8 --check --devices 0.0,1.0
Config:
  scheduler: dynamic
  size: 102400000
  chunksize: 128
  constant: 3.14
  check: yes
  kernel path:examples/tier-2/saxpy.cl
  platform.device list: 0.0 1.0
  static props:
  dynamic chunks: 8
time: 1306
Kernel: saxpy
Runtime init timestamp: 1550921866
Runtime durations:
 initDiscovery: 102 ms.
Device id: 0
Selected platform: Clover
Selected device: AMD Radeon (TM) R7 300 Series (PITCAIRN, DRM 3.27.0, 4.20.10-arch1-1-ARCH, LLVM 7.0.1)
program type: source
kernel: saxpy
works: 2 works_size: 25600000
duration increments:
 init: 0 ms.
 useDiscovery: 0 ms.
 initContext: 0 ms.
 initQueue: 4 ms.
 initBuffers: 0 ms.
 initKernel: 413 ms.
 writeBuffers: 596 ms.
 deviceStart: 0 ms.
 deviceReady: 0 ms.
 deviceRun: 0 ms.
 deviceEnd: 0 ms.
 completeWork: 139 ms.
 total: 1152 ms.
duration offsets from init:
 init: 103 ms.
 useDiscovery: 103 ms.
 initBuffers: 107 ms.
 initKernel: 521 ms.
 writeBuffers: 1117 ms.
 deviceStart: 1117 ms.
 deviceReady: 1117 ms.
 deviceRun: 1117 ms.
 deviceEnd: 1257 ms.
use events: yes
chunks (mOffset+mSize:ts_ms+duration_ms)type-chunks,51200000+12800000:1117+102,76800000+12800000:1219+38,
Device id: 1
Selected platform: AMD Accelerated Parallel Processing
Selected device: AMD FX-8320E Eight-Core Processor
program type: source
kernel: saxpy
works: 6 works_size: 76800000
duration increments:
 init: 0 ms.
 useDiscovery: 0 ms.
 initContext: 0 ms.
 initQueue: 3 ms.
 initBuffers: 0 ms.
 initKernel: 43 ms.
 writeBuffers: 0 ms.
 deviceStart: 0 ms.
 deviceReady: 0 ms.
 deviceRun: 0 ms.
 deviceEnd: 0 ms.
 completeWork: 1153 ms.
 total: 1199 ms.
duration offsets from init:
 init: 102 ms.
 useDiscovery: 103 ms.
 initBuffers: 106 ms.
 initKernel: 150 ms.
 writeBuffers: 150 ms.
 deviceStart: 150 ms.
 deviceReady: 150 ms.
 deviceRun: 150 ms.
 deviceEnd: 1306 ms.
use events: yes
chunks (mOffset+mSize:ts_ms+duration_ms)type-chunks,0+12800000:150+714,12800000+12800000:864+90,25600000+12800000:954+90,38400000+12800000:1045+86,64000000+12800000:1131+88,89600000+12800000:1219+87,
DynamicScheduler:
chunks: 8
duration offsets from init:
 schedulerStart: 0 ms.
 schedulerEnd: 1306 ms.
Success
```

## Contributions

Feel free to contact the maintainer, open issues or do pull requests. We encourage you to request, fix or add functionalities.

## License

Copyright (c) 2018    ATC (University of Cantabria) <nozalr@unican.es>

EngineCL is released under MIT License.

See file LICENSE for full license details.


CHANGELOG.md
# ChangeLog

## v0.4.0 (2019-02-23)

**News**

- Core extracted as a library, supporting Tier-2.
- GitHub publication.

## v0.3.1 (2018-03-20)

**Bugfix**

- Static: needed mutex.
- Dynamic: needed atomic + each device should execute the package given to it.

**Acknowledgements**

- Angélica Dávila (UNIZAR)

## v0.3.0 (2018-03-09)

**Enhancements**

- Named: EngineCL (with `ecl` as namespace)
- Complete functional benchs: saxpy, gaussian, ray, binomial, mandelbrot and nbody.
- Every bench with a OpenCL C++ base case. All but ray with checkers.
- Conditional compilation for logging.
- Improved API for kernel argument (LocalAlloc).
- Output write pattern (relation between gws and index of out buffers).
- Version focused on usability and performance (overhead). Schedulers should be reviewed.
- New reviewed kernels (ray without artifacts).

**Stats**

- C++ LOC: 2219 `src`, 989 `include` (headers)

## v0.2.4 (2018-02-05)

**Enhancements**

- New bench: raytracer.
- Runtime with support of arguments of vector type without buffer initialization (memory allocation for `__local`).
- Buffer simplified and improved with template-based support of data types.
- Flags (`define`) renamed and organized into `config.hpp` file.
- Operations allow blocking reads.

**Stats**

- 2918 C++ lines of code in `src` (852 C/C++ header)

## v0.2.3 (2018-01-29)

**Enhancements**:

- Benchsuite with file reader (source and binary) via `io` compile unit.
- Device supports custom kernels (source and binary).
- Dynamic scheduler with ATOMIC support (`callbacks / get_next_request`).

**News**

- New tool to build kernels: clkernel.

## v0.2.2 (2018-01-22)

**Enhancements**:

- Added support for mOffset in kernel via last args.
- Removed `using namespace std`.
- Added scheduler durations.
- Member vars `size_given` and `size_rem` inside the mutex.

## v0.2.1 (2018-01-17)

**Enhancements**:

- Comments cleaned.
- Inspector created with printActionTypeDuration.
- No warnings in compilation (debug and release).

**News**

- MIT Licensed.
- Clang-format style changed to Mozilla.
- Tested in AMD APU, Intel CPU+IGPU, NVIDIA GPU, AMD CPU+GPU and Xeon Phi.
- Static and Dynamic working.
- HGuided prototype working.
- Makefile improved with clang-tidy and targets refactored.

**Stats**

- 2498 C++ lines of code in `src` (744 C/C++ header)

## v0.2.0 (2018-01-14)

**Enhancements**:

- HGuided load balancer created as prototype.
- Callbacks contain connection with the Work package: `callback` receives the queue index (not the `Device*`).
- Assign, Vecadd and Gaussian benchsuite with an unified structure.

## v0.1.10 (2018-01-03)

**Enhancements**:

- Dynamic load balancer created.
- New library-type API design focused on ease of use.
- Completely new architecture focusing library-runtime reuse.
- CMake organization and library-like building blocks.
- Concurrency mechanism based on callbacks (but easily portable to a direct API-call system).

## v0.1.0 (2017-12-13)

- Static load balancer created.
- Concurrency mechanism based on callbacks.


src/Device.cpp
/**
 * Copyright (c) 2018    ATC (University of Cantabria) <nozalr@unican.es>
 * This file is part of EngineCL which is released under MIT License.
 * See file LICENSE for full license details.
 */
#include "Device.hpp"

#include "Buffer.hpp"
#include "Inspector.hpp"
#include "Runtime.hpp"
#include "Scheduler.hpp"

#define USE_EVENTS 1
// #define USE_EVENTS 0

struct CBData
{
  int queue_index;
  ecl::Device* device;
  CBData(int queue_index_, ecl::Device* device_)
    : queue_index(queue_index_)
    , device(device_)
  {}
};

void CL_CALLBACK
callbackRead(cl_event /*event*/, cl_int /*status*/, void* data)
{
  CBData* cbdata = reinterpret_cast<CBData*>(data);
  ecl::Device* device = cbdata->device;
  ecl::Scheduler* scheduler = device->getScheduler();
#if ECL_SAVE_CHUNKS
  device->saveChunk();
#endif
  device->saveDuration(ecl::ActionType::completeWork);
  scheduler->callback(cbdata->queue_index);
  delete cbdata;
}

namespace ecl {

void
device_thread_func(Device& device)
{
  auto time1 = std::chrono::system_clock::now().time_since_epoch();
  device.init();
  Scheduler* scheduler = device.getScheduler();
  device.saveDuration(ActionType::deviceStart);
  device.saveDurationOffset(ActionType::deviceStart);

  Runtime* runtime = device.getRuntime();
  device.saveDuration(ActionType::deviceReady);
  device.saveDurationOffset(ActionType::deviceReady);
  runtime->notifyReady();
  runtime->notifyAllReady();

  device.waitRun();
  device.saveDuration(ActionType::deviceRun);
  device.saveDurationOffset(ActionType::deviceRun);

  scheduler->requestWork(&device);

  auto cont = true;
  while (cont) {
    device.waitWork();
    auto queue_index = scheduler->getWorkIndex(&device);

    if (queue_index >= 0) { //
      Work work = scheduler->getWork(queue_index);

      device.doWork(work.mOffset, work.mSize, work.mOutWorkitems, work.mOutPositions, queue_index);
    } else {
      device.notifyBarrier();
      cont = false;
    }
  }

  device.saveDuration(ActionType::deviceEnd);
  device.saveDurationOffset(ActionType::deviceEnd);
}

Device::Device(uint selPlatform, uint selDevice)
  : mSelPlatform(selPlatform)
  , mSelDevice(selDevice)
  , mNumArgs(0)
  , mProgramType(ProgramType::Source)
  , mMinMultiplier(1)
{
  mMutexDuration = new mutex();
  mTimeInit = std::chrono::system_clock::now().time_since_epoch();
  mTime = std::chrono::system_clock::now().time_since_epoch();
  mWorks = 0;
  mWorksSize = 0;
  mSemaWork = new Semaphore(1);
  mSemaRun = make_unique<Semaphore>(1);
  mDurationActions.reserve(1024);    // NOTE improve
  mDurationOffsetActions.reserve(8); // the reserve

#if ECL_SAVE_CHUNKS
  mChunks.reserve(160);
#endif
}

Device::~Device()
{
  if (mThread.joinable()) {
    mThread.join();
  }
}

void
Device::printStats()
{
  cout << "Device id: " << getID() << "\n";
  showInfo();
  cout << "program type: ";
  switch (mProgramType) {
    case ProgramType::CustomBinary:
      cout << "custom binary\n";
      break;
    case ProgramType::CustomSource:
      cout << "custom source\n";
      break;
    default:
      cout << "source\n";
  }
  cout << "kernel: " << mKernelStr << "\n";
  cout << "works: " << mWorks << " works_size: " << mWorksSize << "\n";
  size_t acc = 0;
  size_t total = 0;
  cout << "duration increments:\n";
  for (auto& t : mDurationActions) {
    auto d = std::get<0>(t);
    auto action = std::get<1>(t);
    if (action == ActionType::completeWork) {
      acc += d;
    } else {
      Inspector::printActionTypeDuration(action, d);
    }
    total += d;
  }
  cout << " completeWork: " << acc << " ms.\n";
  cout << " total: " << total << " ms.\n";
  cout << "duration offsets from init:\n";
  for (auto& t : mDurationOffsetActions) {
    Inspector::printActionTypeDuration(std::get<1>(t), std::get<0>(t));
  }
#if USE_EVENTS
  cout << "use events: yes\n";
#else
  cout << "use events: no\n";
#endif

#if ECL_SAVE_CHUNKS
  cout << "chunks (mOffset+mSize:ts_ms+duration_ms)";
  cout << "type-chunks,";
  for (auto chunk : mChunks) {
    cout << chunk.offset << "+" << chunk.size << ":" << chunk.ts_ms << "+" << chunk.duration_ms
         << ",";
  }
  cout << "\n";
#endif
}

#if ECL_SAVE_CHUNKS
void
Device::initChunk(size_t offset, size_t size)
{
  mChunk.offset = offset;
  mChunk.size = size;
  auto t2 = std::chrono::system_clock::now().time_since_epoch();
  size_t diff_ms = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - mTimeInit).count();
  mChunk.ts_ms = diff_ms;
  mChunkUnsaved = true;
}

void
Device::saveChunk()
{
  if (mChunkUnsaved) {
    auto t2 = std::chrono::system_clock::now().time_since_epoch();
    size_t diff_ms = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - mTimeInit).count();
    size_t duration_ms = diff_ms - mChunk.ts_ms;
    mChunks.push_back(Chunk(mChunk.offset, mChunk.size, mChunk.ts_ms, duration_ms));
    mChunkUnsaved = false;
  }
}
#endif

void
Device::saveDuration(ActionType action)
{
  lock_guard<mutex> lock(*mMutexDuration);
  auto t2 = std::chrono::system_clock::now().time_since_epoch();
  size_t diff_ms = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - mTime).count();
  mDurationActions.push_back(make_tuple(diff_ms, action));
  mTime = t2;
}
#include <memory>
void
Device::saveDurationOffset(ActionType action)
{
  lock_guard<mutex> lock(*mMutexDuration);
  auto t2 = std::chrono::system_clock::now().time_since_epoch();
  size_t diff_ms = std::chrono::duration_cast<std::chrono::milliseconds>(t2 - mTimeInit).count();
  mDurationOffsetActions.push_back(make_tuple(diff_ms, action));
}

void
Device::setKernelArg(cl_uint index, const uint bytes, ArgType type)
{
  if (type != ArgType::LocalAlloc) {
    throw runtime_error("setLocalArg(uint, uint, ArgType) only admits ArgType::LocalAlloc");
  }
  mArgIndex.push_back(index);
  mArgType.push_back(type);
  mArgBytes.push_back(bytes);
  mArgPtr.push_back(NULL);
  mNumArgs++;
}

void
Device::setKernelArgLocalAlloc(cl_uint index, const uint bytes)
{
  mArgIndex.push_back(index);
  mArgType.push_back(ArgType::LocalAlloc);
  mArgBytes.push_back(bytes);
  mArgPtr.push_back(NULL);
  mNumArgs++;
}

void
Device::setScheduler(Scheduler* scheduler)
{
  mScheduler = scheduler;
}
Scheduler*
Device::getScheduler()
{
  return mScheduler;
}

void
Device::setID(int id)
{
  mId = id;
}
int
Device::getID()
{
  return mId;
}

void
Device::waitRun()
{
  string multiplier("");
  char* val = getenv("MIN_CHUNK_MULTIPLIER");
  if (val != nullptr) {
    multiplier = string(val);
    std::stringstream ss(multiplier);
    std::string item;
    auto i = 0;
    while (std::getline(ss, item, ',')) {
      if (mId == i) {
        mMinMultiplier = stoi(item);
      }
      i++;
    }
  }
  mSemaRun.get()->wait(1);
}

void
Device::notifyRun()
{
  mSemaRun.get()->notify(1);
}

void
Device::waitWork()
{
  mSemaWork->wait(1);
}

void
Device::notifyWork()
{
  mSemaWork->notify(1);
}

string&
Device::getBuffer()
{
  return mInfoBuffer;
}

void
Device::start()
{
  mThread = thread(device_thread_func, std::ref(*this));
}
void
Device::useRuntimeDiscovery()
{
  Runtime* runtime = mRuntime;
  mPlatform = runtime->usePlatformDiscovery(mSelPlatform);
  mDevice = runtime->useDeviceDiscovery(mSelPlatform, mSelDevice);
}

void
Device::setKernel(const string& source, const string& kernel)
{
  if (mProgramType == ProgramType::Source) {
    mProgramSource = source;
  } // else, using custom kernel
  mKernelStr = kernel;
}

void
Device::setKernel(const vector<char>& binary)
{
  mProgramType = ProgramType::CustomBinary;
  mProgramBinary = binary;
}

void
Device::setKernel(const string& source)
{
  mProgramType = ProgramType::CustomSource;
  mProgramSource = source;
}

void
Device::waitEvent()
{
  cl::Event::waitForEvents({ mEnd });
}
void
Device::notifyEvent()
{
  mEnd.setStatus(CL_COMPLETE);
}

void
Device::setBarrier(shared_ptr<Semaphore> barrier)
{
  mBarrier = barrier;
}

void
Device::doWork(size_t poffset, size_t size, uint outWorkitems, uint outPosition, int queueIndex)
{
  if (!size) {
    return callbackRead(nullptr, CL_COMPLETE, this);
  }
  if (mPreviousEvents.size() && mWorks) {
    mPreviousEvents.clear();
  }
  cl::Event evkernel;

  auto gws = size;
  size = outWorkitems * size / outPosition;
  size_t offset = outWorkitems * poffset / outPosition;

  cl_int cl_err;

#if ECL_SAVE_CHUNKS
  initChunk(offset, size);
#endif

#if ECL_KERNEL_GLOBAL_WORK_OFFSET_SUPPORTED == 1
  cl_err = mQueue.enqueueNDRangeKernel(mKernel,
                                       cl::NDRange(poffset),
                                       cl::NDRange(gws),
                                       cl::NDRange(mLws),
                                       &mPreviousEvents,
                                       &evkernel);
#else
  mKernel.setArg(mNumArgs, (uint)poffset);
  cl_err = mQueue.enqueueNDRangeKernel(
    mKernel, cl::NullRange, cl::NDRange(gws), cl::NDRange(mLws), &mPreviousEvents, &evkernel);
#endif
  CL_CHECK_ERROR(cl_err, "enqueue kernel");
#if USE_EVENTS
  cl::Event evread;
  vector<cl::Event> events({ evkernel });
#endif

  auto len = mOutEclBuffers.size();
  for (uint i = 0; i < len; ++i) {
#if USE_EVENTS
    vector<cl::Event> levents = events;
    cl::Event levread;
#endif
    Buffer& b = mOutEclBuffers[i];
    size_t size_bytes = b.byBytes(size);
    auto offset_bytes = b.byBytes(offset);
    cl_err = mQueue.enqueueReadBuffer(mOutBuffers[i],
#if ECL_OPERATION_BLOCKING_READ == 1
                                      CL_TRUE,
#else
                                      CL_FALSE,
#endif
                                      offset_bytes,
                                      size_bytes,
                                      b.dataWithOffset(offset),
#if USE_EVENTS
                                      &levents,
                                      &levread);
    events.push_back(levread);
    evread = levread;
#else
                                      NULL,
                                      NULL);
#endif
    CL_CHECK_ERROR(cl_err, "enqueue read buffer");
  }
  auto cbdata = new CBData(queueIndex, this);

#if ECL_OPERATION_BLOCKING_READ == 1
#if USE_EVENTS
  callbackRead(evread(), CL_COMPLETE, cbdata);
#else
  callbackRead(nullptr, CL_COMPLETE, cbdata);
#endif
#else
  evread.setCallback(CL_COMPLETE, callbackRead, cbdata);
#endif
  mQueue.flush();
  mWorks++;
  mWorksSize += size;
}

void
Device::init()
{
  mTime = std::chrono::system_clock::now().time_since_epoch();
  mInfoBuffer.reserve(128);
  saveDuration(ActionType::init);
  saveDurationOffset(ActionType::init);

  useRuntimeDiscovery();
  saveDuration(ActionType::useDiscovery);
  saveDurationOffset(ActionType::useDiscovery);

  initContext();
  saveDuration(ActionType::initContext);
  initQueue();
  saveDuration(ActionType::initQueue);
  initBuffers();
  saveDuration(ActionType::initBuffers);
  saveDurationOffset(ActionType::initBuffers);
  initKernel();
  saveDuration(ActionType::initKernel);
  saveDurationOffset(ActionType::initKernel);
  initEvents();
  writeBuffers();
  saveDuration(ActionType::writeBuffers);
  saveDurationOffset(ActionType::writeBuffers);
}

void
Device::notifyBarrier()
{
  mBarrier.get()->notify(1);
}

/**
 * \brief Does not check bounds
 */
void
Device::initByIndex(uint selPlatform, uint selDevice)
{
  vector<cl::Platform> platforms;
  cl::Platform::get(&platforms);
  mPlatform = platforms.at(selPlatform);
  vector<cl::Device> devices;
  mPlatform.getDevices(CL_DEVICE_TYPE_ALL, &devices);
  mDevice = devices.at(selDevice);
}

void
Device::initContext()
{
  cl::Context context(mDevice);
  mContext = move(context);
}

void
Device::initQueue()
{
  cl_int cl_err;

  cl::Context& context = mContext;
  cl::Device& device = mDevice;

  cl::CommandQueue queue(context, device, 0, &cl_err);
  CL_CHECK_ERROR(cl_err, "CommandQueue queue");
  mQueue = move(queue);
}

void
Device::initBuffers()
{
  cl_int cl_err = CL_SUCCESS;

  cl_int buffer_in_flags = CL_MEM_READ_WRITE;
  cl_int buffer_out_flags = CL_MEM_READ_WRITE;

  mInBuffers.reserve(mInEclBuffers.size());
  mOutBuffers.reserve(mOutEclBuffers.size());

  auto len = mInEclBuffers.size();
  for (uint i = 0; i < len; ++i) {
    ecl::Buffer& b = mInEclBuffers[i];
    cl::Buffer tmp_buffer(mContext, buffer_in_flags, b.bytes(), NULL);
    CL_CHECK_ERROR(cl_err, "in buffer " + to_string(i));
    mInBuffers.push_back(move(tmp_buffer));
  }

  len = mOutEclBuffers.size();
  for (uint i = 0; i < len; ++i) {
    ecl::Buffer& b = mOutEclBuffers[i];
    cl::Buffer tmp_buffer(mContext, buffer_out_flags, b.bytes(), NULL);
    CL_CHECK_ERROR(cl_err, "out buffer " + to_string(i));
    mOutBuffers.push_back(move(tmp_buffer));
  }
}

void
Device::writeBuffers(bool /* dummy */)
{
  auto len = mInEclBuffers.size();
  mPreviousEvents.reserve(len);
  mPreviousEvents.resize(len);
  for (uint i = 0; i < len; ++i) {
    Buffer& b = mInEclBuffers[i];
    auto data = b.data();
    CL_CHECK_ERROR(mQueue.enqueueWriteBuffer(
      mInBuffers[i], CL_FALSE, 0, b.bytes(), data, NULL, &(mPreviousEvents.data()[i])));
  }
}

void
Device::initKernel()
{
  cl_int cl_err;

  cl::Program::Sources sources;
  cl::Program::Binaries binaries;
  cl::Program program;

  if (mProgramType == ProgramType::CustomBinary) {
    binaries.push_back({ mProgramBinary.data(), mProgramBinary.size() });
#pragma GCC diagnostic ignored "-Wignored-attributes"
    vector<cl_int> status = { -1 };
#pragma GCC diagnostic pop
    program = cl::Program(mContext, { mDevice }, binaries, &status, &cl_err);
    CL_CHECK_ERROR(cl_err, "building program from binary failed for device " + to_string(mId));
  } else {
    sources.push_back({ mProgramSource.c_str(), mProgramSource.length() });
    program = cl::Program(mContext, sources);
  }
  string options;
  options.reserve(32);
  options += "-DECL_KERNEL_GLOBAL_WORK_OFFSET_SUPPORTED=" +
             to_string(ECL_KERNEL_GLOBAL_WORK_OFFSET_SUPPORTED);

  cl_err = program.build({ mDevice }, options.c_str());
  if (cl_err != CL_SUCCESS) {
    cout << " Error building: " << program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(mDevice) << "\n";
    CL_CHECK_ERROR(cl_err);
  }

  cl::Kernel kernel(program, mKernelStr.c_str(), &cl_err);
  CL_CHECK_ERROR(cl_err, "kernel");

  auto len = mArgIndex.size();
  auto unassigned = mInBuffersPtr.size();
  for (uint i = 0; i < len; ++i) {
    auto index = mArgIndex[i];
    auto type = mArgType[i];
    if (type == ArgType::Vector) {
      int pos = -1;
      auto address = mArgPtr[i];
      auto assigned = false;
      if (unassigned > 0) { // usually more in buffers than out
        auto it = find(begin(mInBuffersPtr), end(mInBuffersPtr), address);
        if (it != end(mInBuffersPtr)) {
          pos = distance(mInBuffersPtr.begin(), it);
          cl_err = kernel.setArg(index, mInBuffers[pos]);
          CL_CHECK_ERROR(cl_err, "kernel arg in buffer " + to_string(i));
          unassigned--;
          assigned = true;
        }
      }

      if (pos == -1) {
        auto it = find(begin(mOutBuffersPtr), end(mOutBuffersPtr), address);
        if (it != end(mOutBuffersPtr)) {
          auto pos = distance(mOutBuffersPtr.begin(), it);
          cl_err = kernel.setArg(index, mOutBuffers[pos]);
          CL_CHECK_ERROR(cl_err, "kernel arg out buffer " + to_string(i));
          assigned = true;
        }
      }

      if (!assigned) { // maybe empty space (__local)
        if (mArgBytes[i] > 0) {
          size_t bytes = mArgBytes[i];
          vector<void*>* vptr = reinterpret_cast<vector<void*>*>(mArgPtr[i]);
          void* ptr = vptr->data();
          cl_err = kernel.setArg((cl_uint)i, bytes, ptr);
          CL_CHECK_ERROR(cl_err, "kernel arg " + to_string(i));
        } else {
          throw runtime_error("unknown kernel arg address " + to_string(i));
        }
      }
    } else if (type == ArgType::T) {
      size_t bytes = mArgBytes[i];
      void* ptr = mArgPtr[i];
      cl_err = kernel.setArg((cl_uint)i, bytes, ptr);
      CL_CHECK_ERROR(cl_err, "kernel arg " + to_string(i));
    } else { // ArgType::LocalAlloc
      size_t bytes = mArgBytes[i];
      void* ptr = mArgPtr[i];
      // ptr should be NULL
      cl_err = kernel.setArg((cl_uint)i, bytes, ptr);
      CL_CHECK_ERROR(cl_err, "kernel arg " + to_string(i));
    }
  }

  mKernel = move(kernel);
}

void
Device::setLWS(size_t lws)
{
  mLws = lws;
}

void
Device::setTimeInit(std::chrono::duration<double> timeInit)
{
  mTimeInit = timeInit;
}

void
Device::initEvents()
{
  cl_int cl_err;
  cl::UserEvent end(mContext, &cl_err);
  CL_CHECK_ERROR(cl_err, "user event end");

  mEnd = move(end);
}

void
Device::showInfo()
{
  CL_CHECK_ERROR(mPlatform.getInfo(CL_PLATFORM_NAME, &mInfoBuffer));
  if (mInfoBuffer.size() > 2)
    mInfoBuffer.erase(mInfoBuffer.size() - 1, 1);
  cout << "Selected platform: " << mInfoBuffer << "\n";
  CL_CHECK_ERROR(mDevice.getInfo(CL_DEVICE_NAME, &mInfoBuffer));
  if (mInfoBuffer.size() > 2)
    mInfoBuffer.erase(mInfoBuffer.size() - 1, 1);
  cout << "Selected device: " << mInfoBuffer << "\n";
}

Runtime*
Device::getRuntime()
{
  return mRuntime;
}
void
Device::setRuntime(Runtime* runtime)
{
  mRuntime = runtime;
}

} // namespace ecl
